Failure cluster [ff7a6495...] TestProgressNotify fails when etcd in k/k upgraded to 3.6.2

<img width="1047" height="167" alt="Image" src="https://github.com/user-attachments/assets/8159a0a4-5169-4ae5-bf1a-52d8c14abd92" />


### Failure cluster [ff7a64958c43ff806c16](https://go.k8s.io/triage#ff7a64958c43ff806c16)

##### Error text:
```
Failed
=== RUN   TestProgressNotify
    utils.go:171: time out after waiting 30s on ResultChan
    logger.go:146: 2025-07-10T11:19:52.978Z	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "mux: server closed"}
    logger.go:146: 2025-07-10T11:19:52.978Z	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "http: Server closed"}
    logger.go:146: 2025-07-10T11:19:59.979Z	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "accept tcp 127.0.0.1:39439: use of closed network connection"}
    logger.go:146: 2025-07-10T11:19:59.980Z	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "accept tcp 127.0.0.1:33755: use of closed network connection"}
--- FAIL: TestProgressNotify (37.42s)

```
#### Recent failures:
[7/11/2025, 11:02:08 PM ci-kubernetes-unit-dependencies](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-dependencies/1943868351860183040)
[7/11/2025, 7:02:08 PM ci-kubernetes-unit-dependencies](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-dependencies/1943807952838201344)
[7/11/2025, 3:02:07 PM ci-kubernetes-unit-dependencies](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-dependencies/1943747554797686784)
[7/11/2025, 11:02:08 AM ci-kubernetes-unit-dependencies](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-dependencies/1943687155771510784)
[7/11/2025, 7:02:44 AM ci-kubernetes-unit-dependencies](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-dependencies/1943626757034741760)


/kind failing-test
<!-- If this is a flake, please add: /kind flake -->

<!-- Please assign a SIG using: /sig SIG-NAME -->

--- a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go
@@ -106,9 +106,9 @@ func TestWatchInitializationSignal(t *testing.T) {
 func TestProgressNotify(t *testing.T) {
 	clusterConfig := testserver.NewTestConfig(t)
 	clusterConfig.WatchProgressNotifyInterval = time.Second
-	ctx, store, _ := testSetup(t, withClientConfig(clusterConfig))
+	ctx, store, client := testSetup(t, withClientConfig(clusterConfig))
 
-	storagetesting.RunOptionalTestProgressNotify(ctx, t, store)
+	storagetesting.RunOptionalTestProgressNotify(ctx, t, store, increaseRV(client.Client))
 }
 
 func TestWatchWithUnsafeDelete(t *testing.T) {

--- a/staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go
@@ -585,7 +585,7 @@ func RunTestWatchInitializationSignal(ctx context.Context, t *testing.T, store s
 // Given this feature is currently not explicitly used by higher layers of Kubernetes
 // (it rather is used by wrappers of storage.Interface to implement its functionalities)
 // this test is currently considered optional.
-func RunOptionalTestProgressNotify(ctx context.Context, t *testing.T, store storage.Interface) {
+func RunOptionalTestProgressNotify(ctx context.Context, t *testing.T, store storage.Interface, increaseRV IncreaseRVFunc) {
 	input := &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "name", Namespace: "test-ns"}}
 	key := computePodKey(input)
 	out := &example.Pod{}
@@ -593,6 +593,15 @@ func RunOptionalTestProgressNotify(ctx context.Context, t *testing.T, store stor
 		t.Fatalf("Create failed: %v", err)
 	}
 	validateResourceVersion := resourceVersionNotOlderThan(out.ResourceVersion)
+	// Since etcd v3.6.2 we need to increase RV due to https://github.com/etcd-io/etcd/pull/20241.
+	// We must advance the resource version to ensure that etcd revision progresses past the watch we establish.
+	// As etcd does not send progress notifications for watches on future revisions.
+	//
+	// A Kubernetes watch is exclusive (first event received is after a given RV), which translates
+	// to an inclusive etcd watch at revision+1. Without this increment, if no other writes
+	// have occurred, the watch would be on a future revision, preventing progress
+	// notifications.
+	increaseRV(ctx, t)
 
 	opts := storage.ListOptions{
 		ResourceVersion: out.ResourceVersion,

